lib/IR/DomTreeUpdater.cpp

```diff
@@ -0,0 +1,366 @@
+void DomTreeUpdater::recalculateDomTree(llvm::Function &F) {
+  assert(DT != nullptr && "recalculateDomTree must be called with a DomTree.");
+  if (hasPendDeletedBB() || hasPendingDomTreeUpdates()) {
+    if (hasPendDeletedBB()) {
+      applyPostDomTreeUpdates();
+      forceFlushDeletedBB();
+    }
+    DT->recalculate(F);
+    PendDTUpdateIndex = PendUpdates.size();
+    dropOutOfDateUpdates();
+  }
+}
+// This function should be used when the graph is changed significantly
+// and PDT must be recalculated, previous graph changes should be made.
+void DomTreeUpdater::recalculatePostDomTree(llvm::Function &F) {
+  assert(PDT != nullptr &&
+         "recalculatePostDomTree must be called with a PostDomTree.");
+  if (hasPendDeletedBB() || hasPendingPostDomTreeUpdates()) {
+    if (hasPendDeletedBB()) {
+      applyDomTreeUpdates();
+      forceFlushDeletedBB();
+    }
+    PDT->recalculate(F);
+    PendPDTUpdateIndex = PendUpdates.size();
+    dropOutOfDateUpdates();
+  }
+}
+
+bool DomTreeUpdater::hasPendingUpdates() const {
+  return hasPendingDomTreeUpdates() || hasPendingPostDomTreeUpdates();
+}
+
+bool DomTreeUpdater::hasPendingDomTreeUpdates() const {
+  if (DT == nullptr)
+    return false;
+  return PendUpdates.size() != PendDTUpdateIndex;
+}
+
+bool DomTreeUpdater::hasPendingPostDomTreeUpdates() const {
+  if (PDT == nullptr)
+    return false;
+  return PendUpdates.size() != PendPDTUpdateIndex;
+}
+
+bool DomTreeUpdater::hasPendingDeletedBB(BasicBlock *DelBB) const {
+  if (Ups == UpdateStrategy::Eager || DeletedBBs.empty())
+    return false;
+  return DeletedBBs.count(DelBB) != 0;
+}
+// The DT and PDT require the nodes related to updates
+// are not deleted when update functions are called.
+// So BasicBlock deletions must be pended when the
+// UpdateStrategy is Lazy. When the UpdateStrategy is
+// Eager, the BasicBlock will be deleted immediately.
+void DomTreeUpdater::deleteBB(BasicBlock *DelBB) {
+  validateDeleteBB(DelBB);
+  if (Ups == UpdateStrategy::Lazy)
+    DeletedBBs.insert(DelBB);
+  else
+    DelBB->eraseFromParent();
+}
+
+void DomTreeUpdater::callbackDeleteBB(BasicBlock *DelBB,
+                                      function_ref<void()> Callback) {
+  validateDeleteBB(DelBB);
+  if (Ups == UpdateStrategy::Lazy) {
+    Callbacks.push_back(CallBackOnDeletion(DelBB, Callback));
+    DeletedBBs.insert(DelBB);
+  } else {
+    DelBB->removeFromParent();
+    Callback();
+    delete DelBB;
+  }
+}
+
+void DomTreeUpdater::validateDeleteBB(BasicBlock *DelBB) {
+  assert(DelBB && "Invalid push_back of nullptr DelBB.");
+  assert(pred_empty(DelBB) && "DelBB has one or more predecessors.");
+  // DelBB is unreachable and all its instructions are dead.
+  while (!DelBB->empty()) {
+    Instruction &I = DelBB->back();
+    // Replace used instructions with an arbitrary value (undef).
+    if (!I.use_empty())
+      I.replaceAllUsesWith(llvm::UndefValue::get(I.getType()));
```

Comment:
Maybe this should be made a precondition to the deleteBB function? I think this should be discussed in an upstream review.

Reply:
I agree with you. Honestly, this part of code is moved from DeferredDominance and they make this assumption. :)